src/current/v24.3/troubleshoot-lock-contention.md

```diff
@@ -0,0 +1,395 @@
+
+### Historical queries
+
+In this tutorial's [Example 2](#example-2), Transaction 5 uses a historical query:
+
+~~~ sql
+BEGIN AS OF SYSTEM TIME '-30s';
+SELECT * FROM t2 WHERE k=4; -- historical read
+COMMIT;
+~~~
+
+Consider the following when using [historical queries]({% link {{ page.version.version }}/as-of-system-time.md %}):
+
+- Use historical queries only if the application can use data that is 5 seconds old or older.
+- Historical queries primarily benefit read-only transactions.
+- Historical queries operate below [closed timestamps]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#closed-timestamps) and therefore have perfect concurrency characteristics - they never wait on anything and never block anything.
+- Historical queries have the best possible performance, since they are served by the nearest replica.
+
+### &quot;fail fast&quot; method
+
+In this tutorial's [Example 2](#example-2), Transaction 6 uses a &quot;fail fast&quot; method:
+
+~~~ sql
+BEGIN;
+SELECT * FROM t2 WHERE k=4 FOR UPDATE NOWAIT; -- fail fast write
+UPDATE t2 SET v=4034 WHERE k=4;
+COMMIT;
+~~~
+
+Consider the following when using a ["fail fast" method]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies):
+
+- &quot;Fail fast&quot; could be a reasonable protective measure in the application to handle "hot update key" situations, for example, when an application needs to be able to handle an arbitrary large surge of updates on the same key.
+- The most direct method of "failing fast" is using pessimistic locking with [SELECT FOR UPDATE … NOWAIT]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies). It can reduce or prevent failures late in a transaction's life (e.g. at the `COMMIT` time), by returning an error early in a contention situation if a row cannot be locked immediately.
+- A more &quot;buffered fail fast&quot; approach would be to control the maximum length of a lock wait-queue that requests are willing to enter and wait in, with the cluster setting  [`kv.lock_table.maximum_lock_wait_queue_length`](https://github.com/cockroachdb/cockroach/pull/66146) (default: `0`). It can provide some level of quality-of-service with a response time predictability in a severe per-key contention. If set to a non-zero value and an existing lock wait-queue is already equal to or exceeding this length, requests will be rejected eagerly instead of entering the queue and waiting.
```

Comment:
this is a non-public cluster setting that is undocumented

if we're not going to the trouble of flipping that cluster setting to public (a 1-line code change) we should definitely not be adding it to the docs

we should also DEFINITELY not be linking to a github PR as the reference docs for something that has pretty deep consequences for cluster operation

Reply:
removed

Reply:
removed